drivers/frequency/adf4382/adf4382.c

```diff
@@ -0,0 +1,1303 @@
+#include "no_os_error.h"
+#include "no_os_print_log.h"
+
+/* Charge pump current values expressed in uA */
+static const int adf4382_ci_ua[] = {
+	700,
+	900,
+	1100,
+	1300,
+	1400,
+	1800,
+	2200,
+	2500,
+	2900,
+	3600,
+	4300,
+	5000,
+	5800,
+	7200,
+	8600,
+	11100
+};
+
+/**
+ * @brief Writes data to ADF4382 over SPI.
+ * @param dev	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param data 	   - Data value to write.
+ * @return 	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_write(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t data)
+{
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
+	if (!dev)
+		return -EINVAL;
+
+	cmd = ADF4382_SPI_WRITE_CMD | reg_addr;
+	if (dev->spi_desc->bit_order) {
+		buff[0] = no_os_bit_swap_constant_8(cmd & 0xFF);
+		buff[1] = no_os_bit_swap_constant_8(cmd >> 8);
+		buff[2] = no_os_bit_swap_constant_8(data);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = data;
+	}
+
+	return no_os_spi_write_and_read(dev->spi_desc, buff,
+					ADF4382_BUFF_SIZE_BYTES);
+}
+
+/**
+ * @brief Reads data from ADF4382 over SPI.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param data	   - Data read from the device.
+ * @return	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_read(struct adf4382_dev *dev, uint16_t reg_addr, uint8_t *data)
+{
+	int ret;
+	uint8_t buff[ADF4382_BUFF_SIZE_BYTES];
+	uint16_t cmd;
+
+	if (!dev)
+		return -EINVAL;
+
+	cmd = ADF4382_SPI_READ_CMD | reg_addr;
+	if (dev->spi_desc->bit_order) {
+		buff[0] = no_os_bit_swap_constant_8(cmd & 0xFF);
+		buff[1] = no_os_bit_swap_constant_8(cmd >> 8);
+		buff[2] = no_os_bit_swap_constant_8(ADF4382_SPI_DUMMY_DATA);
+	} else {
+		buff[0] = cmd >> 8;
+		buff[1] = cmd & 0xFF;
+		buff[2] = ADF4382_SPI_DUMMY_DATA;
+	}
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buff,
+				       ADF4382_BUFF_SIZE_BYTES);
+	if(ret)
+		return ret;
+
+	*data = buff[2];
+
+	return 0;
+}
+
+/**
+ * @brief Updates the values of the ADF4382 register.
+ * @param dev 	   - The device structure.
+ * @param reg_addr - The register address.
+ * @param mask 	   - Bits to be updated.
+ * @param data 	   - Update value for the mask.
+ * @return	   - 0 in case of success or negative error code otherwise.
+ */
+int adf4382_spi_update_bits(struct adf4382_dev *dev, uint16_t reg_addr,
+			    uint8_t mask, uint8_t data)
+{
+	uint8_t tmp, orig;
+	int ret;
+
+	ret = adf4382_spi_read(dev, reg_addr, &orig);
+	if (ret)
+		return ret;
+
+	tmp = orig & ~mask;
+	tmp |= data & mask;
+
+	if (tmp != orig)
+		return adf4382_spi_write(dev, reg_addr, tmp);
+
+	return 0;
+}
+
+/**
+ * @brief Will output on the terminal the values of all the ADF4382 registers.
+ * @param dev 	- The device structure.
+ * @return 	- 0 in case of success or negative error code.
+ */
+int adf4382_reg_dump(struct adf4382_dev *dev)
+{
+	uint8_t val;
+	uint8_t ret;
```

Comment:
if ret can take negative values, then use `int` type. Check throughout the code, since this is occuring multiple times.